### PR TITLE
Rebuild devices in Frame Looping

### PR DESCRIPTION
**Pull request by purvisa-at-google-com**

### CODE REVIEW

File: gapis/api/vulkan/frame_loop.go [hunk 1]
```diff
@@ -1360,6 +1414,273 @@ func (f *frameLoop) resetResources(ctx context.Context, stateBuilder *stateBuild
...
+	for _, deviceObject := range GetState(f.loopStartState).Devices().All() {
+
+		physicalDevice := deviceObject.PhysicalDevice()
+		physicalDeviceObject := GetState(f.loopStartState).PhysicalDevices().All()[physicalDevice]
+		instance := physicalDeviceObject.Instance()
+
+		if _, ok := f.instanceToCreate[instance]; ok {
+			f.deviceToDestroy[deviceObject.VulkanHandle()] = true
+			f.deviceToCreate[deviceObject.VulkanHandle()] = true
+		}
```
Comment (RenfengLiu) [inline gapis/api/vulkan/frame_loop.go]: Do we need to handle surface here?

File: gapis/api/vulkan/frame_loop.go [hunk 2]
```diff
@@ -1360,6 +1414,273 @@ func (f *frameLoop) resetResources(ctx context.Context, stateBuilder *stateBuild
...
+		device := memoryObject.Device()
+
+		if _, ok := f.deviceToCreate[device]; ok {
+			f.memoryToFree[memoryObject.VulkanHandle()] = true
+			f.memoryToAllocate[memoryObject.VulkanHandle()] = true
+			// TODO: @renfengliu What do we need to do here with memoryToUnmap/memoryToMap/mappedAddress?
+		}
+	}
+
+	for _, bufferObject := range GetState(f.loopStartState).Buffers().All() {
```
Comment (RenfengLiu) [inline gapis/api/vulkan/frame_loop.go]: How about bufferview?

File: gapis/api/vulkan/frame_loop.go [hunk 3]
```diff
@@ -1360,6 +1414,273 @@ func (f *frameLoop) resetResources(ctx context.Context, stateBuilder *stateBuild
...
+		device := bufferObject.Device()
+
+		if _, ok := f.deviceToCreate[device]; ok {
+			f.bufferToDestroy[bufferObject.VulkanHandle()] = true
+			f.bufferToCreate[bufferObject.VulkanHandle()] = true
+			// TODO: @renfengliu What do we need to do here with bufferChanged/bufferToRestore?
+		}
+	}
+
+	for _, imageObject := range GetState(f.loopStartState).Images().All() {
```
Comment (RenfengLiu) [inline gapis/api/vulkan/frame_loop.go]: How about imageview

File: gapis/api/vulkan/frame_loop.go (hunk 2 above)
Comment (purvisa-at-google-com) [reply gapis/api/vulkan/frame_loop.go]: BufferView and ImageView should be taken care of inductively by rebuilding the buffers and images they are built from.

File: gapis/api/vulkan/frame_loop.go (hunk 3 above)
Comment (purvisa-at-google-com) [reply gapis/api/vulkan/frame_loop.go]: See BufferView above!

File: gapis/api/vulkan/frame_loop.go [hunk 4]
```diff
@@ -1360,6 +1414,273 @@ func (f *frameLoop) resetResources(ctx context.Context, stateBuilder *stateBuild
...
+		device := imageObject.Device()
+
+		if _, ok := f.deviceToCreate[device]; ok {
+			f.imageToDestroy[imageObject.VulkanHandle()] = true
+			f.imageToCreate[imageObject.VulkanHandle()] = true
+			// TODO: @renfengliu What do we need to do here with imageChanged/imageToRestore?
+		}
+	}
+
+	for _, samplerYcbcrObject := range GetState(f.loopStartState).SamplerYcbcrConversions().All() {
```
Comment (RenfengLiu) [inline gapis/api/vulkan/frame_loop.go]: This duplicates line 1507, maybe just sampler ?

File: gapis/api/vulkan/frame_loop.go (hunk 1 above)
Comment (purvisa-at-google-com) [reply gapis/api/vulkan/frame_loop.go]: Good catch. That was in code review while I was doing this work, and I based the list of things to rebuild off the code as checked in to master.

File: gapis/api/vulkan/frame_loop.go (hunk 4 above)
Comment (purvisa-at-google-com) [reply gapis/api/vulkan/frame_loop.go]: Another good catch. I will remove one of the two identical loops.

**Pull request merged**